pkg/webhook/resources/virtualmachine/validator.go

```diff
@@ -0,0 +1,79 @@
+package virtualmachine
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubevirtv1 "kubevirt.io/client-go/api/v1"
+
+	"github.com/harvester/harvester/pkg/util"
+	werror "github.com/harvester/harvester/pkg/webhook/error"
+	"github.com/harvester/harvester/pkg/webhook/types"
+)
+
+func NewValidator() types.Validator {
+	return &vmValidator{}
+}
+
+type vmValidator struct {
+	types.DefaultValidator
+}
+
+func (v *vmValidator) Resource() types.Resource {
+	return types.Resource{
+		Name:       "virtualmachines",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   kubevirtv1.SchemeGroupVersion.Group,
+		APIVersion: kubevirtv1.SchemeGroupVersion.Version,
+		ObjectType: &kubevirtv1.VirtualMachine{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+			admissionregv1.Update,
+		},
+	}
+}
+
+func (v *vmValidator) Create(request *types.Request, newObj runtime.Object) error {
+	vm := newObj.(*kubevirtv1.VirtualMachine)
+
+	if err := v.checkVolumeClaimTemplatesAnnotation(vm); err != nil {
+		message := fmt.Sprintf("the volumeClaimTemplates annotaion is invalid: %v", err)
+		return werror.NewInvalidError(message, "metadata.annotations")
+	}
+	return nil
+}
```

Comment:
The volumeClaimTemplate annotation works like `statefulset.spec.volumeClaimTemplates`. It creates PVCs if they do not exist. We don't clean up created PVCs when the VM is deleted. We don't block VM creation if the PVCs already exist.

Reply:
If the user accidentally creates the `volumeClaimTemplates` using the same name, then two VMs will be mounted to the same PVC and it may fail to start and be hard to debug as there is no obvious event or errors.
And maybe we just can log the error events of the VM instead of adding the annotation validation.

Reply:
Addressed in https://github.com/harvester/harvester/pull/1066/commits/9476ffda434f268762ef4685db4af7ae10c985ba